### PR TITLE
win,fs: get fstat when no permission

### PR DESCRIPTION
**Pull request by huseyinacacak-janea**
If file permissions are restricted, calling `CreateFileW` returns a permission error. This avoids getting the basic file information. 
This PR fixes this problem, by calling `FindFirstFileW` to get basic file information. 

Fixes: https://github.com/libuv/libuv/issues/1980
Fixes: https://github.com/libuv/libuv/issues/3267
Refs: https://github.com/nodejs/node/issues/35853

### CODE REVIEW

File: src/win/fs.c [hunk 1]
```diff
@@ -1971,8 +1971,51 @@ INLINE static DWORD fs__stat_impl_from_path(WCHAR* path,
...
-  if (handle == INVALID_HANDLE_VALUE)
-    return GetLastError();
+  if (handle == INVALID_HANDLE_VALUE) {
+    DWORD error = GetLastError();
+    int sys_error = uv_translate_sys_error(error);
+    WIN32_FIND_DATAW findFile_data;
+    HANDLE find_handle;
+    FILE_STAT_BASIC_INFORMATION stat_info;
+
+    if (sys_error != UV_EBUSY && sys_error != UV_EPERM) {
```
Comment (saghul) [inline src/win/fs.c:1981]: Wouldn't it make more sense to check the native Windows error here?

File: src/win/fs.c [hunk 2]
```diff
@@ -1971,8 +1971,51 @@ INLINE static DWORD fs__stat_impl_from_path(WCHAR* path,
...
+      return error;
+    }
+
+    // If the file isn't open, try to get the file statistics using FindFirstFileW
+    find_handle = FindFirstFileW(path, &findFile_data);
+
+    if (find_handle == INVALID_HANDLE_VALUE) {
+      return GetLastError();
```
Comment (saghul) [inline src/win/fs.c:1989]: Should we return this error or the original one?

File: src/win/fs.c [hunk 3]
```diff
@@ -1971,8 +1971,51 @@ INLINE static DWORD fs__stat_impl_from_path(WCHAR* path,
...
+    int sys_error = uv_translate_sys_error(error);
+    WIN32_FIND_DATAW findFile_data;
+    HANDLE find_handle;
+    FILE_STAT_BASIC_INFORMATION stat_info;
+
+    if (sys_error != UV_EBUSY && sys_error != UV_EPERM) {
+      return error;
+    }
+
+    // If the file isn't open, try to get the file statistics using FindFirstFileW
```
Comment (saghul) [inline src/win/fs.c:1985]: Nit: use /* */ style of comments.

File: src/win/fs.c [hunk 4]
```diff
@@ -1971,8 +1971,51 @@ INLINE static DWORD fs__stat_impl_from_path(WCHAR* path,
...
+      return GetLastError();
+    }
+
+    stat_info.DeviceType = 0;
+    stat_info.FileAttributes = findFile_data.dwFileAttributes;
+    stat_info.NumberOfLinks = 0;
+    stat_info.FileId.QuadPart = 0;
+    stat_info.AllocationSize.QuadPart = 0;
+    stat_info.VolumeSerialNumber.QuadPart = 0;
+    stat_info.CreationTime.QuadPart =
```
Comment (saghul) [inline src/win/fs.c:1998]: Maybe use a macro to avoid the repetition for the 4 fields?

**Comment by vtjnash** [comment]
This is fixed on most file systems by https://github.com/libuv/libuv/pull/4327 already. See also https://github.com/libuv/libuv/pull/3268#discussion_r1007238152 which was a review of a past attempt at implementing this.

File: src/win/fs.c (hunk 1 above)
Comment (huseyinacacak-janea) [reply src/win/fs.c:1981]: I used the translated errors because if a new error is mapped to these 2 in the future, it might be worth trying this `FindFristFileW` for the new ones. Someone can add a new map in the header file but can easily forget to add it here. 
However, if you want me to copy all those native Windows errors, I can do so.

**Comment by huseyinacacak-janea** [comment]
> This is fixed on most file systems by #4327 already. See also [#3268 (comment)](https://github.com/libuv/libuv/pull/3268#discussion_r1007238152) which was a review of a past attempt at implementing this.

Thanks for the review. 
The PR https://github.com/libuv/libuv/pull/4327 seems to be not fixing the issue. The test in this PR fails without `FindFirstFileW()`.
Sorry, I couldn't quite understand what you mean by https://github.com/libuv/libuv/pull/3268#discussion_r1007238152. Do you want me to use `NtQueryDirectoryFile` directly?

**Comment by vtjnash** [comment]
The CI workers are likely not using a kernel where this feature is implemented yet (https://github.com/libuv/libuv/pull/4327#issuecomment-1976833275)

**Comment by vtjnash** [comment]
> Sorry, I couldn't quite understand what you mean by https://github.com/libuv/libuv/pull/3268#discussion_r1007238152. Do you want me to use NtQueryDirectoryFile directly?

It just meant I didn't directly review this, but was pointing out that you may have the same implementation issues as the past attempt:
  - NtQueryDirectoryFile is more feature-full (and faster), so is likely a better choice, as well as being pretty much a drop-in replacement
  - This actually implements lstat, but it claims to implement stat, so that may need to be fixed

**Comment by huseyinacacak-janea** [comment]
I see your point. Yes, https://github.com/libuv/libuv/pull/4327 fixes for the new Windows versions. But, I think it would be better to fix this issue for the older versions too.

As you suggested, I wanted to use `NtQueryDirectoryFile` to fix this issue and see the performance improvement over `FindFirstFile`.  AFAIK, `FindFirstFile` uses `NtOpenFile` and `NtQueryDirectoryFile` under the hood. Therefore, I tried to use both of them. However, I'm constantly getting `ENOENT` errors when calling `NtOpenFile`. Am I using these functions incorrectly? I appreciate your help.

```patch
diff --git a/src/win/fs.c b/src/win/fs.c
index 1f255408..47dd732c 100644
--- a/src/win/fs.c
+++ b/src/win/fs.c
@@ -1940,6 +1940,30 @@ INLINE static void fs__stat_prepare_path(WCHAR* pathw) {
   }
 }
 
+#define InitializeObjectAttributes(p, n, a, r, s) { \
+    (p)->Length = sizeof(OBJECT_ATTRIBUTES);        \
+    (p)->RootDirectory = r;                         \
+    (p)->Attributes = a;                            \
+    (p)->ObjectName = n;                            \
+    (p)->SecurityDescriptor = s;                    \
+    (p)->SecurityQualityOfService = NULL;           \
+}
+VOID RtlInitUnicodeString(
+    PUNICODE_STRING DestinationString,
+    PCWSTR SourceString
+) {
+    if (DestinationString != NULL) {
+        if (SourceString != NULL) {
+            DestinationString->Buffer = (PWSTR)SourceString;
+            DestinationString->Length = (USHORT)wcslen(SourceString) * sizeof(WCHAR);
+            DestinationString->MaximumLength = DestinationString->Length + sizeof(WCHAR);
+        } else {
+            DestinationString->Buffer = NULL;
+            DestinationString->Length = 0;
+            DestinationString->MaximumLength = 0;
+        }
+    }
+}
 
 INLINE static DWORD fs__stat_impl_from_path(WCHAR* path,
                                             int do_lstat,
@@ -1974,21 +1998,42 @@ INLINE static DWORD fs__stat_impl_from_path(WCHAR* path,
   if (handle == INVALID_HANDLE_VALUE) {
     DWORD error = GetLastError();
     int sys_error = uv_translate_sys_error(error);
-    WIN32_FIND_DATAW findFile_data;
+    //WIN32_FIND_DATAW findFile_data;
     HANDLE find_handle;
     FILE_STAT_BASIC_INFORMATION stat_info;
+    FILE_DIRECTORY_INFORMATION dir_info;
+    IO_STATUS_BLOCK ioStatusBlock;
+    NTSTATUS status;
+    PCWSTR PartName;
+    UNICODE_STRING NtName;
+    OBJECT_ATTRIBUTES objAttributes;
+    RTL_RELATIVE_NAME RelativeName;
 
     if (sys_error != UV_EBUSY && sys_error != UV_EPERM) {
       return error;
     }
 
-    // If the file isn't open, try to get the file statistics using FindFirstFileW
-    find_handle = FindFirstFileW(path, &findFile_data);
+    RtlInitUnicodeString(&NtName, path);
+    InitializeObjectAttributes(&objAttributes, &NtName, OBJ_CASE_INSENSITIVE, NULL, NULL);
+
+    // Attempt to open the directory with FILE_LIST_DIRECTORY access
+    status = pNtOpenFile(&find_handle, FILE_LIST_DIRECTORY, &objAttributes, &ioStatusBlock,
+                         FILE_SHARE_READ | FILE_SHARE_WRITE | FILE_SHARE_DELETE,
+                         /*FILE_DIRECTORY_FILE*/0x00000001);
 
-    if (find_handle == INVALID_HANDLE_VALUE) {
+    if (!NT_SUCCESS(status)) {
+      SetLastError(pRtlNtStatusToDosError(status));
       return GetLastError();
     }
 
+    /* If the file isn't open, try to get the file statistics using FindFirstFileW */
+    // Query the directory
+    status = pNtQueryDirectoryFile(handle, NULL, NULL, NULL, &ioStatusBlock, &dir_info, sizeof(FILE_FULL_DIR_INFO), 
+                                   FileFullDirectoryInformation, TRUE, NULL, TRUE);
+    if (!NT_SUCCESS(status)) {
+      return error;
+    }
+
     stat_info.DeviceType = 0;
     stat_info.FileAttributes = findFile_data.dwFileAttributes;
     stat_info.NumberOfLinks = 0;

```

**Comment by vtjnash** [comment]
NtOpenFile is very difficult to use, but it is basically just an inferior version of CreateFile

**Comment by vtjnash** [comment]
You should be able to use CreateFile on the directory (with the appropriate flags) instead

**Comment by huseyinacacak-janea** [comment]
Is there anything else I can do to help this PR move forward?

File: src/win/fs.c [hunk 5]
```diff
@@ -1940,6 +1940,166 @@ INLINE static void fs__stat_prepare_path(WCHAR* pathw) {
   }
 }
 
+INLINE static int fs__stat_split_path(const WCHAR* filename, WCHAR** dir,
```
Comment (saghul) [inline src/win/fs.c:1943]: Is this code basically a copy of `uv__split_path` in fs-event.c? If so, it should be factored out elsewhere, not copied here...

File: src/win/fs.c [hunk 6]
```diff
@@ -1940,6 +1940,166 @@ INLINE static void fs__stat_prepare_path(WCHAR* pathw) {
...
+        break;
+      }
+      SetLastError(pRtlNtStatusToDosError(status));
+      ret_error = GetLastError();
+      goto cleanup;
+    }
+
+    pdir_info = buf;
+
+    /* Check files one by one */
```
Comment (saghul) [inline src/win/fs.c:2060]: Isn't this racey?

File: src/win/fs.c [hunk 7]
```diff
@@ -1940,6 +1940,166 @@ INLINE static void fs__stat_prepare_path(WCHAR* pathw) {
...
+    status = pNtQueryDirectoryFile(handle,
+                                   NULL,
+                                   0,
+                                   0,
+                                   &ioStatusBlock,
+                                   buf,
+                                   sizeof(buf),
+                                   FileFullDirectoryInformation,
+                                   FALSE,
+                                   NULL,
```
Comment (vtjnash) [inline src/win/fs.c:2046]: It might be much easier to pass the FileName argument here, and then just asserting there is only one?

File: src/win/fs.c (hunk 6 above)
Comment (vtjnash) [reply src/win/fs.c:2060]: Yes, the `stat` syscall is race-y

File: src/win/fs.c [hunk 8]
```diff
@@ -1940,6 +1940,166 @@ INLINE static void fs__stat_prepare_path(WCHAR* pathw) {
...
+  stat_info.AllocationSize.QuadPart = pdir_info->AllocationSize.QuadPart;
+
+  fs__stat_assign_statbuf(statbuf, stat_info, do_lstat);
+
+cleanup:
+  if (path_filename) {
+    uv__free(path_filename);
+    path_filename = NULL;
+  }
+  FindClose(handle);
```
Comment (vtjnash) [inline src/win/fs.c:2100]:
```suggestion
  CloseHandle(handle);
```

File: src/win/fs-event.c [hunk 9]
```diff
@@ -80,7 +80,7 @@ static void uv__relative_path(const WCHAR* filename,
   (*relpath)[relpathlen] = L'\0';
 }
 
-static int uv__split_path(const WCHAR* filename, WCHAR** dir,
+int uv__fs_event_split_path(const WCHAR* filename, WCHAR** dir,
```
Comment (saghul) [inline src/win/fs-event.c:83]: This function is not really related to fs events, I think it'd be better named `uv__fs_split_path` and moved to fs.c.

File: src/win/fs.c (hunk 5 above)
Comment (huseyinacacak-janea) [reply src/win/fs.c]: Yes, it was a copy of it. I've added the definition of the original implementation to `internal.h` .  I hope I put it in the right place.

File: src/win/fs.c (hunk 8 above)
Comment (huseyinacacak-janea) [reply src/win/fs.c]: Thanks. fixed.

File: src/win/fs.c (hunk 7 above)
Comment (huseyinacacak-janea) [reply src/win/fs.c:1991]: I tried your suggestion but got the same error as in the issue (UV__EACCES).

File: src/win/fs.c (hunk 6 above)
Comment (saghul) [reply src/win/fs.c:2005]: I suppose it's a best-effort case here, and the look will end if there are no more entries, so I guess it's ok...

**Review by saghul — APPROVED** [review]
LGTM, but I'd like @vtjnash to also take a final look.

File: src/win/fs.c [hunk 10]
```diff
@@ -1941,6 +1941,110 @@ INLINE static void fs__stat_prepare_path(WCHAR* pathw) {
...
+    status = pNtQueryDirectoryFile(handle,
+                                   NULL,
+                                   0,
+                                   0,
+                                   &ioStatusBlock,
+                                   buf,
+                                   sizeof(buf),
+                                   FileFullDirectoryInformation,
+                                   FALSE,
+                                   NULL,
```
Comment (vtjnash) [inline src/win/fs.c:1991]: The FileName should probably be passed here. Not sure how my comment ended up so far away on a different line

File: src/win/fs.c [hunk 11]
```diff
@@ -1941,6 +1941,110 @@ INLINE static void fs__stat_prepare_path(WCHAR* pathw) {
...
+  NTSTATUS status;
+  WCHAR buf[32 * 1024];
+  WCHAR* path_dirpath = NULL;
+  WCHAR* path_filename = NULL;
+  BOOL ret_val = FALSE;
+  DWORD ret_error = 0;
+
+  if (uv__fs_split_path(path, &path_dirpath, &path_filename) != 0) {
+    ret_error = GetLastError();
+    goto cleanup;
```
Comment (vtjnash) [inline src/win/fs.c:1960]: HANDLE looks uninitialized here, but looks like it gets passed to CloseHandle

File: test/test-fs.c [hunk 12]
```diff
@@ -4507,6 +4507,62 @@ TEST_IMPL(fs_open_readonly_acl) {
...
+        goto acl_cleanup;
+    }
+
+    /* Read file stats */
+    r = uv_fs_stat(NULL, &req, filename, NULL);
+    if (r != 0) {
+        goto acl_cleanup;
+    }
+
+    fflush(stdout);
```
Comment (vtjnash) [inline test/test-fs.c:4552]: Old debugging leftover?

File: test/test-fs.c [hunk 13]
```diff
@@ -4507,6 +4507,62 @@ TEST_IMPL(fs_open_readonly_acl) {
...
+    uv_fs_req_cleanup(&req);
+
+ acl_cleanup:
+    /* Cleanup */
+    call_icacls("icacls %s /reset", filename);
+    uv_fs_unlink(NULL, &unlink_req, filename, NULL);
+    uv_fs_req_cleanup(&unlink_req);
+    unlink(filename);
```
Comment (vtjnash) [inline test/test-fs.c:4560]: Use uv_fs_remove for correct Unicode handling here

Comment (vtjnash) [inline test/test-fs.c:4560]: We possibly should be using uv_fs_remove for correct Unicode handling here, but looks like this is consistent with the other tests and should be okay

File: src/win/fs.c (hunk 10 above)
Comment (huseyinacacak-janea) [reply src/win/fs.c:1991]: I tried your suggestion but got the same error as in the issue (UV__EACCES).

File: src/win/fs.c (hunk 11 above)
Comment (huseyinacacak-janea) [reply src/win/fs.c:1960]: Fixed.

File: test/test-fs.c (hunk 12 above)
Comment (huseyinacacak-janea) [reply test/test-fs.c]: Yes, deleted it.

File: test/test-fs.c (hunk 13 above)
Comment (huseyinacacak-janea) [reply test/test-fs.c:4559]: As you said, I used it for consistency.

**Review by vtjnash — COMMENTED** [review]
Spent some time investigating, testing, and finishing this today, of which https://github.com/libuv/libuv/pull/4566 is my updates to this PR now. The EACCES was very suspicious sounding to me, and turns out it was a segfault being misreported there. Fixing the API call makes it work as expected.

**Comment by huseyinacacak-janea** [comment]
Closing in favor of https://github.com/libuv/libuv/pull/4566.
